pkg/xds/generator/ingress_generator.go

```diff
@@ -129,6 +157,49 @@ func (_ IngressGenerator) destinations(
 		}
 	}
 
+	if len(ingressProxy.PolicyResources[meshhttproute_api.MeshHTTPRouteType].GetItems()) > 0 {
+		// We need to add a destination to route any service to any instance of
+		// that service
+		matchAllTags := tags.Tags{
+			mesh_proto.ServiceTag: mesh_proto.MatchAllTag,
+		}
+		matchAllDestinations := destinations[mesh_proto.MatchAllTag]
+		foundAllServicesDestination := slices.IndexFunc(matchAllDestinations, func(tagsElem tags.Tags) bool {
+			return reflect.DeepEqual(tagsElem, matchAllTags)
+		}) > -1
+		if !foundAllServicesDestination {
+			matchAllDestinations = append(matchAllDestinations, matchAllTags)
+		}
+		destinations[mesh_proto.MatchAllTag] = matchAllDestinations
+	}
+
+	// Note that we're not merging these resources, but that's OK because the
+	// set of destinations after merging is a subset of the set we get here by
+	// iterating through them.
+	for _, route := range ingressProxy.PolicyResources[meshhttproute_api.MeshHTTPRouteType].(*meshhttproute_api.MeshHTTPRouteResourceList).Items {
+		for _, to := range route.Spec.To {
+			toTags, ok := tagsFromTargetRef(to.TargetRef)
+			if !ok {
+				continue
+			}
+
+			service := toTags[mesh_proto.ServiceTag]
```

Comment:
I don't think it's correct to take `service` from to-targetRef, it should be taken from `backendRef`. For example:
```yaml
spec:
  targetRef:
    kind: Mesh
  to:
    - targetRef:
        kind: MeshService
        name: legacy-payments
      rules:
        - matches:
            - path:
                type: Prefix
                value: /
          default:
            backendRefs:
              - kind: MeshService
                name: new-payments
                tags:
                  version: v1
                weight: 90
              - kind: MeshService
                name: new-payments
                tags:
                  version: v2
                weight: 10                 
```
so in `destinations` you want to have:
```yaml
new-payments: # not 'legacy-payments'
  - kuma.io/service: new-payments
    version: v1
  - kuma.io/service: new-payments
    version: v2
```
at least this is how TrafficRoute works with Ingress

Reply:
It's only used in the case that `BackendRefs` is empty. Otherwise we use the `backendRefs` below.

Reply:
yeah I think that makes sense because at this point the ZoneIngress will never see `legacy-payments`.

Reply:
Also, I hope we'll be able to implement this one https://github.com/kumahq/kuma/issues/5827

Reply:
I want to add this case to the e2e test as well

Reply:
Added to e2e and fixed, WDYT?

Reply:
Resolving this